### PR TITLE
fix: [#1311] schema validation

### LINKED ISSUE #1311: Service and data source schema validation
Example schema validation failure:

`2024-04-24 08:59:47 | app.transform.utils.validate | WARNING | service - input schema validation failure. Column types missmatch. Difference: {'multimedia_urls': {'ACTUAL': 'array<string>', 'EXPECTED': 'array<struct<name:string,url:string>>'}, 'upstream_id': {'ACTUAL': 'string', 'EXPECTED': 'bigint'}, 'use_cases_urls': {'ACTUAL': 'array<string>', 'EXPECTED': 'array<struct<name:string,url:string>>'}} | validate.py/validate_schema [39]`

In the handling of `services` and `data sources`, we encounter an issue where `multimedia_urls` and `use_cases_url` can either be blank lists `[]` or lists populated with dictionaries that contain `name` and `url` keys, e.g., `[{'name': 'examplename', 'url': 'https://www.exampleurl.com'}]`. When these fields are blank, the schema validation incorrectly interprets them as `array<string>` instead of the expected `array<struct<name:string,url:string>>`. This discrepancy leads to schema validation failures, necessitating a review of whether the schema should be adjusted to accept both formats or if the MP should modify how these fields are transmitted to the transformer.

Additionally, there is a similar issue with `upstream_ud`, which is typically a `bigint`. However, it can also be transmitted as an empty field, which is then incorrectly treated by the schema as a `string`.

**Resolution Suggestion**:
We need to decide whether to adjust the schema to accommodate these variations or to alter the data transmission approach from the MP to ensure consistency with the expected schema types.

### PR DESCRIPTION
**Pull request by wiktorflorian**
Closes #1311 

In some cases, services and data sources may have two input types for fields: `multimedia_urls`, `upstream_id`, and `use_cases_urls`. This is due to the fact that fields can be left blank. The input schemas have been modified to be dictionaries with lists of strings as values. Additionally, the validation functions have been updated to accommodate dictionaries with lists. This approach enables us to validate the input type against the specified input schema types.

Additionally, the mappings were updated and the value `black .` was used.

### CODE REVIEW

**Review by Michal-Kolomanski — APPROVED** [review]
typing adjustments :D

File: transform/app/services/solr/validate/schema/validate.py [hunk 1]
```diff
@@ -74,26 +94,37 @@ def get_column_type(column):
...
-    schema = {col: get_column_type(col) for col in df.columns}
-    return schema
+    return {col: get_column_type(col) for col in df.columns}
+
+
+def is_type_match(actual_type: str, expected_type: Union[List[str], str]) -> bool:
+    """Check if the actual type matches the expected type(s)"""
+    if isinstance(expected_type, list):
+        return (
+            actual_type in expected_type or actual_type == "void" or actual_type == ""
```
[the remark this answers is not in the document]
Comment (wiktorflorian) [reply transform/app/services/solr/validate/schema/validate.py:104]: I attempted various approaches and forgot to delete it.

File: transform/app/services/solr/validate/schema/validate.py [hunk 2]
```diff
@@ -74,26 +94,37 @@ def get_column_type(column):
...
-    for actual_sch, (key, expected_sch) in zip_longest(
-        actual_schema, expected_schema.items()
-    ):
-        if actual_sch not in (expected_sch, "void"):
-            differences[key] = {"ACTUAL": actual_sch, "EXPECTED": expected_sch}
-
+    for col, actual_sch in zip_longest(columns, actual_schema):
+        expected_sch = expected_schema[col]
+        if not is_type_match(actual_sch, expected_sch):
+            logger.warning(f"Mismatch found in column '{col}'")
```
[the remark this answers is not in the document]
Comment (wiktorflorian) [reply transform/app/services/solr/validate/schema/validate.py:123]: You are correct. I added additional logging to identify the point at which validations fail, but neglected to remove it.

File: transform/app/services/solr/validate/schema/validate.py [hunk 3]
```diff
@@ -20,49 +20,69 @@ def validate_schema(
...
+
+def validate_column_types(
+    df: DataFrame,
+    expected_schema: Dict[str, Union[List[str], str]],
+    collection: str,
+    source: Literal["input", "output"],
+) -> None:
+    """Validate that column types match"""
+    actual_types = [column.dataType.simpleString() for column in df.schema.fields]
+    differences = print_diff_schema(df.columns, actual_types, expected_schema)
```
[the remark this answers is not in the document]
Comment (wiktorflorian) [reply transform/app/services/solr/validate/schema/validate.py:49]: Will `get_schmea_differences` be ok?

**Pull request merged**